### PR TITLE
:dart: test for io and gmsh, fix #4

### LINKED ISSUE #4: Add unit tests
* Using the `catch` framework add unit tests to all classes

### PR DESCRIPTION
**Pull request by ezrayst**

### CODE REVIEW

File: tests/gmsh_test.cc [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+
+//! \brief Check that IO will store right values
+TEST_CASE("GMSH is checked in 3D", "[GMSH][3D]") {
+
+  //! 3D Test of IO Class
+  const unsigned DIM = 3;
+  const double tolerance = 1.E-12;
+
+  auto mesh = std::unique_ptr<Mesh<3, 4>> (new GMSH<3, 4>());
+  mesh->read_mesh("../bin/test_3d.msh");
```
Comment (kks32) [inline tests/gmsh_test.cc]: Please don't hardcode paths.

File: tests/gmsh_test.cc [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+  REQUIRE(mesh->material_points().size() == 4);
+
+  //! Check id
+  REQUIRE(mesh->material_points().at(0)->id() == 1);
+  REQUIRE(mesh->material_points().at(1)->id() == 2);
+  REQUIRE(mesh->material_points().at(2)->id() == 3);
+  REQUIRE(mesh->material_points().at(3)->id() == 4);
+
+  //! Check coordinates
+  REQUIRE(mesh->material_points().at(0)->coordinates().at(0) == Approx(0.5).epsilon(tolerance));
```
Comment (kks32) [inline tests/gmsh_test.cc]: This is not a unit test, you are trying to do a regression test here!

**Review by kks32 — CHANGES_REQUESTED** [review]

File: include/gmsh.tcc [hunk 3]
```diff
@@ -80,6 +81,7 @@ void GMSH<Tdim, Tvertices>::read_vertices(std::ifstream& file) {
       this->vertices_.insert(std::make_pair(vertid, vertex));
     }
   }
+<<<<<<< HEAD
```
Comment (kks32) [inline include/gmsh.tcc]: Check merge, this seems like there is a merge issue.

File: include/gmsh.tcc [hunk 4]
```diff
@@ -246,7 +251,8 @@ void GMSH<Tdim, Tvertices>::compute_stresses() {
   const double k0 = 0.5;
   const double conv_factor = 10;
 
-  double max_height = std::numeric_limits<double>::min();;
+  double max_height = std::numeric_limits<double>::min();
+  ;
```
Comment (kks32) [inline include/gmsh.tcc]: Remove this `;`

**Review by kks32 — CHANGES_REQUESTED** [review]

File: include/gmsh.tcc (hunk 3 above)
Comment (ezrayst) [reply include/gmsh.tcc]: checked it in the next commit

Comment (ezrayst) [reply include/gmsh.tcc]: actually I had it in the next commit already sometime ago

File: include/gmsh.tcc [hunk 5]
```diff
@@ -4,137 +4,151 @@
...
-      if (line.find('#') == std::string::npos && line != "") {
-        //! Coordinates of vertex
-        std::array<double, Tdim> vertex;
-
-        istream >> vertid;
-        istream >> vertex.at(0) >> vertex.at(1) >> vertex.at(2);
-        vertices_.insert(std::make_pair(vertid, vertex));
-      }
+      istream >> vertid;
+      istream >> vertex.at(0) >> vertex.at(1) >> vertex.at(2);
```
Comment (kks32) [inline include/gmsh.tcc]: Is this true for 2D and 3D?

File: include/gmsh.tcc [hunk 6]
```diff
@@ -4,137 +4,151 @@
...
+      this->vertices_.insert(std::make_pair(vertid, vertex));
     }
-    infile.close();
   }
-  std::cout << "Number of Vertices: " << vertices_.size() << '\n';
+
+  this->nvertices_ = vertices_.size();
+
+  //! Check that the number of vertices are correct
+  if (nvertices_ != nvertices) {
```
Comment (kks32) [inline include/gmsh.tcc]: Remove `{` `}` for a single line loops/conditions.

File: include/gmsh.tcc [hunk 7]
```diff
@@ -4,137 +4,151 @@
...
+    std::cout << "Error in the code, number of vertices do not match.\n";
```
Comment (kks32) [inline include/gmsh.tcc]: Rephrase error message, `error in the code` isn't very helpful.

File: include/gmsh.tcc [hunk 8]
```diff
@@ -4,137 +4,151 @@
...
+  std::cout << "Number of Vertices: " << nvertices_ << '\n';
 }
 
 //! Read GMSH elements
 //! \tparam Tdim Dimension
 //! \tparam Tvertices Number of vertices in element
 //! \param[in] filename Input mesh filename
 template <unsigned Tdim, unsigned Tvertices>
-void GMSH<Tdim, Tvertices>::read_elements(const std::string& filename) {
+void GMSH<Tdim, Tvertices>::read_elements(std::ifstream& file) {
```
Comment (kks32) [inline include/gmsh.tcc]: Function definition doesn't match the comments.

File: include/gmsh.tcc [hunk 9]
```diff
@@ -4,137 +4,151 @@
...
-    //! Skip past vertices section
-    for (int i = 0; i < nvertices; ++i) {
-      std::getline(infile, line);
-    }
-    //! Skip past element headers
-    for (unsigned l = 0; l < 2; ++l) {
-      std::getline(infile, line);
-    }
+  //! specify element type 4 = tetrahedral
+  const unsigned element_type = 4;
```
Comment (kks32) [inline include/gmsh.tcc]: I thought we use `quadrilateral` or `hexahedron`.

File: include/gmsh.tcc [hunk 10]
```diff
@@ -4,137 +4,151 @@
...
-              elementarray.at(2) >> elementarray.at(3);
-
-          elements_.insert(std::make_pair(elementid, elementarray));
-        }
+  //! Iterate through all elements in the file
+  for (unsigned i = 0; i < nelements; ++i) {
+    std::getline(file, line);
+    std::istringstream istream(line);
+    if (line.find('#') == std::string::npos && line != "") {
+      istream >> elementid;
```
Comment (kks32) [inline include/gmsh.tcc]: Can we have these in a single `istream >> elementid >> elementtype >> ...`

File: include/gmsh.tcc [hunk 11]
```diff
@@ -4,137 +4,151 @@
...
+      istream >> elementtype;
+      istream >> elementry;
+      istream >> physical;
+      istream >> elementry;
+
+      //! If element type not == to specified Tvertices, skip element
```
Comment (kks32) [inline include/gmsh.tcc]: Rewrite as `not equals to`

**Review by kks32 — CHANGES_REQUESTED** [review]
This PR tries to fix 2 issues: Refactor GMSH #15 to include search using `keywords` and Unit tests #4. Could you split this so that they two separate PRs. The unit tests still needs work. 

Please keep commit messages descriptive, don't use `Fix Krishna's comments`, write what you did `Added keyword search` instead.

File: include/gmsh.tcc (hunk 5 above)
Comment (ezrayst) [reply include/gmsh.tcc]: in 2D, we will have the z coordinate as 0 right? so this is a general statement, no?

File: include/gmsh.tcc (hunk 9 above)
Comment (ezrayst) [reply include/gmsh.tcc]: in `mesh` file, we use number 4 or 8 I think

File: include/gmsh.tcc (hunk 5 above)
Comment (kks32) [reply include/gmsh.tcc]: Is that true for GMSH?

File: include/gmsh.tcc (hunk 9 above)
Comment (kks32) [reply include/gmsh.tcc]: yes, but `tetraheron` is not an element we use.

File: include/gmsh.tcc (hunk 8 above)
Comment (ezrayst) [reply include/gmsh.tcc]: can you be more specific? is it because there is no directory for the filename?

**Comment by ezrayst** [comment]
So this PR would be in relation to issue #15 only right, and not unit tests?

File: include/gmsh.tcc (hunk 9 above)
Comment (ezrayst) [reply include/gmsh.tcc]: Understood!

**Comment by kks32** [comment]
You need to move the tests to a different branch and just commit changes related to #15

**Comment by ezrayst** [comment]
What is the best way to do it? Just revert my push and then push again without those files regarding unit-test?

**Comment by kks32** [comment]
Create a new branch and move relevant files to there and create a new PR. Or, create a new branch, move tests to that branch and delete it from this PR.

**Comment by ezrayst** [comment]
This pull request is now only for unit-test

File: CMakeLists.txt [hunk 12]
```diff
@@ -51,6 +51,8 @@ add_executable(mpm_point_generator ${mpm_point_gen_src})
 SET(test_src
   ${PROJECT_SOURCE_DIR}/tests/test.cc
   ${PROJECT_SOURCE_DIR}/tests/point_test.cc
+  ${PROJECT_SOURCE_DIR}/tests/io_test.cc
+  ${PROJECT_SOURCE_DIR}/tests/gmsh_test.cc
```
Comment (kks32) [inline CMakeLists.txt:55]: This is a regression test and not a unit test and should therefore be built separately.

**Review by kks32 — CHANGES_REQUESTED** [review]

**Comment by ezrayst** [comment]
The test for `gmsh`, I still don't know how to test it apart from reading a file. Could @kks32 explain more on how to do this? I am getting quite confused.

**Comment by kks32** [comment]
This is fine, just make the test part a separate executable.

**Comment by kks32** [comment]
Can we have just the test case here? The other changes have been made elsewhere?

**Comment by ezrayst** [comment]
I think it should just be the tests... not sure why the others are showing as a change where the change has been made previously...

**Comment by ezrayst** [comment]
Note the test fails because we have not fixed #34 yet... But the rest should run as it is

**Comment by kks32** [comment]
Rebase and have only changes with respect to the test. The test should be a separate executable in CMake, and not part of the unit test.

**Comment by ezrayst** [comment]
I have rebased by the way, and I think it's best to have a new PR on this issue as rebasing might be quite convoluted...

You mean you want the `gmsh` test to be separate from unit test?

**Comment by kks32** [comment]
Closing this as #35 supersedes this!

**Pull request closed without merging**